crates/subspace-runtime-primitives/src/extension.rs

```diff
@@ -123,14 +147,41 @@ where
         _inherited_implication: &impl Encode,
         _source: TransactionSource,
     ) -> ValidateResult<Self::Val, RuntimeCallFor<Runtime>> {
-        let validity = if origin.as_system_origin_signer().is_some() {
-            Self::do_validate_signed(call)?
+        let (validity, maybe_calls) = if origin.as_system_origin_signer().is_some() {
+            Self::do_validate_signed(call).map(|(valid, calls)| (valid, Some(calls)))?
         } else {
-            ValidTransaction::default()
+            (ValidTransaction::default(), None)
         };
 
-        Ok((validity, (), origin))
+        Ok((validity, maybe_calls, origin))
     }
 
-    impl_tx_ext_default!(RuntimeCallFor<Runtime>; prepare);
+    fn prepare(
+        self,
+        val: Self::Val,
+        _origin: &DispatchOriginOf<RuntimeCallFor<Runtime>>,
+        _call: &RuntimeCallFor<Runtime>,
+        _info: &DispatchInfoOf<RuntimeCallFor<Runtime>>,
+        _len: usize,
+    ) -> Result<Self::Pre, TransactionValidityError> {
+        let assigned_weight = Self::get_weights(MAXIMUM_NUMBER_OF_CALLS);
+        match val {
+            None => Ok(Pre::Refund(assigned_weight)),
```

Comment:
There is a db read in `Runtime::is_balance_transferable` which applies to all extrinsics and should be refunded here.